LLama/Model/ModelCache.cs

```diff
@@ -0,0 +1,238 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using LLama.Common;
+using LLama.Native;
+
+namespace LLama.Model;
+
+/// <inheritdoc />
+public class ModelCache : IModelCache
+{
+    private bool _disposed = false;
+
+    /// <summary>
+    /// Support model type files
+    /// </summary>
+    public static readonly string[] ExpectedModelFileTypes = [
+        ".gguf"
+    ];
+
+    // keys are directories, values are applicable models
+    private readonly Dictionary<string, IEnumerable<ModelFileMetadata>> _availableModels = [];
+
+    // model id/alias, to loaded model
+    private readonly Dictionary<string, SafeLlamaModelHandle> _loadedModelCache = [];
+
+    /// <summary>
+    /// Create a new model manager that seeds available models from the given directory list
+    /// </summary>
+    /// <param name="directories"></param>
+    public ModelCache(string[] directories)
+    {
+        GetModelsFromDirectories(directories);
+    }
+
+    /// <inheritdoc />
+    public IEnumerable<ModelFileMetadata> ModelFileList
+        => _availableModels.SelectMany(x => x.Value);
+
+    /// <inheritdoc />
+    public IEnumerable<string> ModelDirectories
+        => _availableModels.Keys;
+
+    #region Directories
+    private void GetModelsFromDirectories(params string[] dirs)
+    {
+        foreach (var dir in dirs)
+        {
+            var fullDirectoryPath = Path.GetFullPath(dir);
+
+            if (!Directory.Exists(fullDirectoryPath))
+            {
+                Trace.TraceError($"Model directory '{fullDirectoryPath}' does not exist");
+                continue;
+            }
+
+            if (_availableModels.ContainsKey(fullDirectoryPath))
+            {
+                Trace.TraceWarning($"Model directory '{fullDirectoryPath}' already probed");
+                continue;
+            }
+
+            // find models in current dir that are of expected type
+            List<ModelFileMetadata> directoryModelFiles = [];
+            foreach (var file in Directory.EnumerateFiles(fullDirectoryPath))
+            {
+                if (!ExpectedModelFileTypes.Contains(Path.GetExtension(file)))
+                {
+                    continue;
+                }
+
+                // expected model file
+                var fi = new FileInfo(file);
```

Comment:
Please check whether this file is a symbolic link. If so, we should get the byte size from its target file.

If I'm not misunderstanding it, the semantic here is the real model size?

Reply:
Checking for symbolic links would be good but it's non trivial to handle it cross platform. Can this be deferred and addressed separately? 

It's just a reflection of the size of the actual data on disk. I'm not sure if that maps directly to the model's underlying size or not. I think a future enhancement would be to read part(s) of GGUF header to get some more metadata about the model like name, architecture, etc but I kept it to just the file system info for now.

Reply:
It's okay to keep it now but `FileInfo.LinkTarget` is a cross-platform API to get the target file path with .NET version > NetStandard2.0

Reply:
yeah, I found that too and didnt want to deal with 2.0 case 😄

Reply:
```suggestion
                var fi = new FileInfo(file);
                // TODO: check whether the file is a symbolic link
```

OK, please leave a comment here.